src/coprocessor/endpoint.rs

```diff
@@ -485,33 +486,107 @@ impl<E: Engine> Endpoint<E> {
     #[inline]
     pub fn parse_and_handle_unary_request(
         &self,
-        req: coppb::Request,
+        mut req: coppb::Request,
         peer: Option<String>,
     ) -> impl Future<Output = MemoryTraceGuard<coppb::Response>> {
         let tracker = GLOBAL_TRACKERS.insert(::tracker::Tracker::new(RequestInfo::new(
             req.get_context(),
             RequestType::Unknown,
             req.start_ts,
         )));
+        let batch_task_num = req.tasks.len();
+        let mut batch_fut_results = Vec::with_capacity(batch_task_num + 1);
```

Comment:
I suggest to extract all the batch part to a new function. And I think it's not necessary to combine the batch tasks with the original one. 

Then, we don't need the extra allocation if there is no batch tasks. The future that joins all batch tasks also process the results into `batch_responses`. I think it will be clearer.

Reply:
My initial thought is that combining all the futures of original and batch tasks and waiting for them together so the latency would be `max(req_latency)`, but the logic also becomes complex. If the batched tasks are processed and waited separately the latency would be increased, especially when the original task is executing slowly. Any suggestion about this?

Reply:
It's still possible to execute them concurrently.

One is the original task `fut1` and the other is the batched tasks `fut2` created by `join_all`. Then, use the `futures::join` macro to get both the results.